flink-connectors/flink-connector-elasticsearch6/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/table/Elasticsearch6DynamicSinkFactory.java

```diff
@@ -0,0 +1,152 @@
+import static org.apache.flink.streaming.connectors.elasticsearch.table.ElasticsearchOptions.BULK_FLUSH_INTERVAL_OPTION;
+import static org.apache.flink.streaming.connectors.elasticsearch.table.ElasticsearchOptions.BULK_FLUSH_MAX_ACTIONS_OPTION;
+import static org.apache.flink.streaming.connectors.elasticsearch.table.ElasticsearchOptions.CONNECTION_MAX_RETRY_TIMEOUT_OPTION;
+import static org.apache.flink.streaming.connectors.elasticsearch.table.ElasticsearchOptions.CONNECTION_PATH_PREFIX;
+import static org.apache.flink.streaming.connectors.elasticsearch.table.ElasticsearchOptions.DOCUMENT_TYPE_OPTION;
+import static org.apache.flink.streaming.connectors.elasticsearch.table.ElasticsearchOptions.FAILURE_HANDLER_OPTION;
+import static org.apache.flink.streaming.connectors.elasticsearch.table.ElasticsearchOptions.FLUSH_ON_CHECKPOINT_OPTION;
+import static org.apache.flink.streaming.connectors.elasticsearch.table.ElasticsearchOptions.FORMAT_OPTION;
+import static org.apache.flink.streaming.connectors.elasticsearch.table.ElasticsearchOptions.HOSTS_OPTION;
+import static org.apache.flink.streaming.connectors.elasticsearch.table.ElasticsearchOptions.INDEX_OPTION;
+import static org.apache.flink.streaming.connectors.elasticsearch.table.ElasticsearchOptions.KEY_DELIMITER_OPTION;
+
+/**
+ * A {@link DynamicTableSinkFactory} for discovering {@link Elasticsearch6DynamicSink}.
+ */
+@Internal
+public class Elasticsearch6DynamicSinkFactory implements DynamicTableSinkFactory {
+	private static final Set<ConfigOption<?>> requiredOptions = Stream.of(
+		HOSTS_OPTION,
+		INDEX_OPTION,
+		DOCUMENT_TYPE_OPTION
+	).collect(Collectors.toSet());
+	private static final Set<ConfigOption<?>> optionalOptions = Stream.of(
+		KEY_DELIMITER_OPTION,
+		FAILURE_HANDLER_OPTION,
+		FLUSH_ON_CHECKPOINT_OPTION,
+		BULK_FLASH_MAX_SIZE_OPTION,
+		BULK_FLUSH_MAX_ACTIONS_OPTION,
+		BULK_FLUSH_INTERVAL_OPTION,
+		BULK_FLUSH_BACKOFF_TYPE_OPTION,
+		BULK_FLUSH_BACKOFF_MAX_RETRIES_OPTION,
+		BULK_FLUSH_BACKOFF_DELAY_OPTION,
+		CONNECTION_MAX_RETRY_TIMEOUT_OPTION,
+		CONNECTION_PATH_PREFIX,
+		FORMAT_OPTION
+	).collect(Collectors.toSet());
+
+	@Override
+	public DynamicTableSink createDynamicTableSink(Context context) {
+		ElasticsearchValidationUtils.validatePrimaryKey(context.getCatalogTable().getSchema());
+		final FactoryUtil.TableFactoryHelper helper = FactoryUtil.createTableFactoryHelper(this, context);
+
+		final SinkFormat<SerializationSchema<RowData>> format = helper.discoverSinkFormat(
+			SerializationFormatFactory.class,
+			FORMAT_OPTION);
+
+		helper.validate();
+		Configuration configuration = new Configuration();
+		context.getCatalogTable()
+			.getOptions()
+			.forEach(configuration::setString);
+		Elasticsearch6Configuration config = new Elasticsearch6Configuration(configuration, context.getClassLoader());
+
+		validate(config, configuration);
+
+		return new Elasticsearch6DynamicSink(
+			format,
+			config,
+			context.getCatalogTable().getSchema());
```

Comment:
It's a little tricky here. We should use `TableSchemaUtils.getPhysicalSchema(context.getCatalogTable().getSchema())` here to get rid of computed columns. However `TableSchemaUtils.getPhysicalSchema` will remove primary key information too. Maybe we should keep primary key in `TableSchemaUtils.getPhysicalSchema` and use it here.

Reply:
I admit I missed that bit. 

The schema is used for two purposes:
* generating dynamic index -> it should work on the computed columns as well imo
* computing id based on the primary key -> primary key cannot be defined on a computed column so it is not a problem

For the primary keys I think it's fine because it primary keys cannot be defined on computed columns. It is a problem for the dynamic index though. I will pass the physical schema and keys separately for now.

Reply:
I think it's also a problem for the key extract, because we are using the index to extract key. The RowData consumed by sink do not contain computed column. Maybe we should add test for this.

Reply:
You are right. BTW I think the `FileSystemTableSink` have the same problem.

Reply:
nice discussion